File: c++/0045. K Closest Points to Origin.cpp
// time = O(nlgk)
// space = 

using pi = pair<int, vector<int>>;

class Compare {
public:
    bool operator() (pi &pair1, pi &pair2) {
        return pair1.first < pair2.first;
    }
};

class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        priority_queue<pi, vector<pi>, Compare> pq;
        for (auto &point: points) {
            int x = point.front(), y = point.back();
            pq.push({x*x+y*y, point});
            if (pq.size() > k) pq.pop();
        }
        
        points.clear();
        for (int i = 0; i< k; i++) {
            points.push_back(pq.top().second);
            pq.pop();
        }
        return points;
    }
};

File: 0045_k_closest_points_to_origin.py
# time: O(nlgk)
# space: O(k)
from heapq import heappop, heappush
from typing import List
from math import sqrt
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        def distance(x, y):
            return sqrt(x*x + y*y)
        heap_arr = []
        for point in points:
            heappush(heap_arr, (-distance(*point), point[0], point[1]))
            if len(heap_arr) > k:
                heappop(heap_arr)
        return [i[1:] for i in heap_arr]